---
params:
  sample: Example
  unfiltered_sce: !r scpcaTools:::sim_sce()
  filtered_sce: NULL
  date: !r Sys.Date()

title: "`r glue::glue('ScPCA QC report for {params$sample}')`"
author: "CCDL"
date: "`r params$date`"
output: 
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    number_sections: true
    code_download: true
---

```{r setup, message = FALSE, echo = FALSE}
# knitr options
knitr::opts_chunk$set(
  echo = FALSE
)

library(SingleCellExperiment)
library(dplyr)
library(ggplot2)

# Set standard ggplot theme
theme_set(theme_bw())
```

```{r sce_setup}
# save some typing later
sample <- params$sample
unfiltered_sce <- params$unfiltered_sce
filtered_sce <- params$filtered_sce

has_filtered <- !is.null(filtered_sce)

# if there is no filtered sce, use the unfiltered for both
if (!has_filtered){
  filtered_sce <- unfiltered_sce
}
# add cell stats if missing
if (is.null(unfiltered_sce$sum)){
  unfiltered_sce <- scater::addPerCellQC(unfiltered_sce)
}
```

# Introduction

# Processing Information for `r sample`

## Raw Sample Metrics

```{r }
# extract sce metadata containing processing information as table
unfiltered_meta <- metadata(unfiltered_sce) 

sample_information <- tibble::tibble(
  "Sample id" = sample,
  "Tech version" = unfiltered_meta$tech_version,
  "Number of reads sequenced" = format(unfiltered_meta$total_reads, big.mark = ',', scientific = FALSE),
  "Number of mapped reads" = format(unfiltered_meta$mapped_reads, big.mark = ',', scientific = FALSE),
  "Number of cells reported by alevin-fry" = format(unfiltered_meta$af_num_cells, big.mark = ',', scientific = FALSE)
) %>%
  t()

# make table with sample information
knitr::kable(sample_information, "simple")
```

## Pre-Processing Information

```{r }
processing_info <- tibble::tibble(
  "Salmon version" = unfiltered_meta$salmon_version,
  "Alevin-fry version" = unfiltered_meta$alevinfry_version,
  "Transcriptome index" = unfiltered_meta$reference_index,
  "Filtering method" = unfiltered_meta$af_permit_type,
  "Resolution" = unfiltered_meta$af_resolution, 
  "Transcripts included" = dplyr::case_when(
      unfiltered_meta$transcript_type == "spliced" ~ "Spliced only",
      unfiltered_meta$transcript_type == "unspliced" ~ "Spliced and unspliced" )
  ) %>%
  t()


# make table with processing information
knitr::kable(processing_info, "simple")
```

# `r sample` Experiment Summary 

This sample has `r ncol(sce)` cells, assayed for `r nrow(sce)` genes.

```{asis, echo = has_filtered}
After filtering, there are `r ncol(filtered_sce)` cells.
```

# Session Info
```{r session_info}
sessioninfo::session_info()
```


